### PR TITLE
s390x: update some regalloc metadata to remove use of `reg_mod`.

### PR DESCRIPTION
**Pull request by cfallin**
This is a step toward ultimately removing modify-operands, which along
with removal of pinned vregs, lets us move to a completely
constraint-based and fully-SSA regalloc input and get some nice
advantages eventually.

There are still a few uses of `mod` operands and pinned vregs remaining,
especially around the "regpair" abstraction. Those proved to be a bit
trickier to update though, so will have to be done separately.

### CODE REVIEW

**Comment by cfallin** [comment]
cc @uweigand to review?

I also spent the past day trying to go further in [this branch](https://github.com/cfallin/wasmtime/tree/s390x-ra2-semantics-2), and managed to clean up a lot more, but there are some panics from undefined regs there so I don't have it quite right. @uweigand if you have time to poke at this more I'd very much appreciate it (but it's not urgent at all!). My long-term goal is to push the regalloc input toward fully-SSA (no mods, no multiple defs) and fully constraint-based (no pinned vregs) input, which allows for more flexibility in managing copies and spills, and makes for a more efficient solver generally.

File: cranelift/codegen/src/isa/s390x/inst.isle
Comment (uweigand) [reply cranelift/codegen/src/isa/s390x/inst.isle:2716]: In fact, the whole `same_reg` logic can probably go away then, as the assertion that the two register are actually tied is now made during `emit` anyway.

File: cranelift/filetests/filetests/isa/s390x/arithmetic.clif
Comment (cfallin) [reply cranelift/filetests/filetests/isa/s390x/arithmetic.clif]: This is a great idea actually; thanks! It nicely solves the issue of showing all info when viewing pre-regalloc VCode, but showing machine code as it really is once regalloc occurs. I've made this change, and I plan to adopt it to the mod operands in the other backends as well.

File: cranelift/codegen/src/isa/s390x/inst/emit.rs
Comment (cfallin) [reply cranelift/codegen/src/isa/s390x/inst/emit.rs]: Updated to use `rn`'s value, thanks. I actually lean slightly toward keeping these fields named `ri`, to make it clear that they are artificial instruction fields, the "input" side of the dest reg, rather than a real `rn` field (this also made it easier to grep for things when updating tests just now!). But I'm happy to alter the field name as well if you feel strongly about this.

File: cranelift/codegen/src/isa/s390x/inst.isle
Comment (cfallin) [reply cranelift/codegen/src/isa/s390x/inst.isle]: Done!

**Comment by cfallin** [comment]
Thanks @uweigand! I've updated based on feedback (and, importantly, reverted the 2-to-3-arg change in assembly printing). Thanks for looking further at the followup patch as well; I will pick that up and try to finish it next week, most likely.

File: cranelift/filetests/filetests/isa/s390x/bitops.clif
Comment (uweigand) [inline cranelift/filetests/filetests/isa/s390x/bitops.clif:42]: These two `lgr` look new - this is why the new code is two instructions longer than the old code.  Not sure if this is something that could still be improved in regalloc, or if this is just one of those random changes ...   In any case, not a big deal, I just wanted to point it out.

**Review by uweigand — APPROVED** [review]
LGTM now.   Note the inline comment about one minor regalloc regression, but that doesn't block this PR.

File: cranelift/filetests/filetests/isa/s390x/bitops.clif
Comment (cfallin) [reply cranelift/filetests/filetests/isa/s390x/bitops.clif:42]: I briefly looked but it was nothing really obvious. It's possible that my changes in bytecodealliance/regalloc2#74 might help a bit, but I'm not sure; let's see if it reverts back once I update tests there with this merged :-)

**Comment by cfallin** [comment]
Ah, I think this needs an r+ from someone with write access to the repo -- anyone want to give a rubberstamp on top of Ulrich's review above?

**Review by alexcrichton — APPROVED** [review]

**Pull request merged**
